### PR TITLE
feat(kuma-cp): CircuitBreaker proposal

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary

An initial version of the proposal for the new policy "CircuitBreaker"

### CODE REVIEW

File: docs/proposals/circuit-breaker.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  "max_requests": "{...}",
+  "max_retries": "{...}",
+  "retry_budget": "{...}",
+  "track_remaining": "...",
+  "max_connection_pools": "{...}"
+}
+```
+So circuit breaker in Envoy is a bunch of thresholds which prevents service from opening more new connections and create more new requests than configured. Also, it could limit the number of retries that happen in parallel. 
+
+That functionality of Envoy doesn't allow implementing design pattern described above. In order to make it work there is another useful Envoy feature called `cluster.OutlierDetection`:
```
Comment (jakubdyszkiewicz) [inline docs/proposals/circuit-breaker.md:32]: We use outlier detection already for passive healthchecks https://kuma.io/docs/0.5.0/policies/health-check/ (check `ClusterWithHealthChecks` method). I think it makes sense to add more settings to passive healthchecks or get rid of passive HC and create Circuit Breaker.

File: docs/proposals/circuit-breaker.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+{
+  "priority": "...",
+  "max_connections": "{...}",
+  "max_pending_requests": "{...}",
+  "max_requests": "{...}",
+  "max_retries": "{...}",
+  "retry_budget": "{...}",
+  "track_remaining": "...",
+  "max_connection_pools": "{...}"
+}
```
Comment (jakubdyszkiewicz) [inline docs/proposals/circuit-breaker.md:28]: This configuration is also important and can be a blocker when adopting Kuma. Especially `max_connections`

**Comment by lobkovilya** [comment]
@subnetmarco looks like we have several ways to implement CircuitBreaker:
1. Get rid of passive HealthCheck and have all kinds of circuit breaking in one Policy:

```yaml
type: CircuitBreaker
mesh: default
name: cb-1
selector:
  - match:
      service: backend 
      region: aws
      version: 3
conf:
  maxConnections: 3
  maxPendingRequest: 10
  ...
  interval: 1s
  baseEjectionTime: 30s
  maxEjectionPercent: 20
  splitExternalAndLocalErrors: false
  detectors:
    errors:
      total: 20
      gateway: 10
      local: 7
    standardDeviation:
      requestVolume: 10
      minimumHosts: 5
      factor: 1.9
    failure:
      requestVolume: 10
      minimumHosts: 5
      threshold: 85
```
This approach closer to design patter called "Circuit breaker".

2. Leave HealthCheck policy as it is. Enhance passive HealthCheck with additional parameters for outlier detection. Create CircuitBreaker policy that easily maps to Envoy's `cluster.CircuitBreakers`:
```yaml
type: CircuitBreaker
mesh: default
name: cb-1
selector:
  - match:
      service: backend 
      region: aws
      version: 3
conf:
  maxConnections: 3
  maxPendingRequest: 10
  maxRequests: 10
  maxRetries: 3
```
Write documentation page "Circuit Breakers" where we will show both passive HealthCheck and CircuitBreaker.
This approach probably closer to Envoy users.

Do we have any notion what users expect from Kuma when it comes to Circuit breakers?

**Comment by subnetmarco** [comment]
I suggest that we add the `CircuitBreaker` resource and remove passive healthchecking from the `Healthcheck` resource.

**Pull request merged**